Convert OTA requestor cluster to the code-driven style

harimau-qirex opened the pull request:
#### Summary

This converts the OTA requestor cluster to use the code-driven style, reusing the existing OTA requestor logic as much as possible.

The original generated code compatibility layer in ota-requestor-server.h/cpp has been replaced with the following:
* The attribute storage for all attributes other than `UpdatePossible` was moved to `DefaultOTARequestor` directly. `UpdatePossible` was added to the new data model class `OTARequestorCluster`.
* Events are now handled by the `OTARequestorEventHandler` interface that `OTARequestorCluster` implements. `DefaultOTARequestor` keeps a registry of event handlers.
* Commands are still forwarded to `DefaultOTARequestor`, now originating from `OTARequestorCluster`.

Miscellaneous changes:
* Support for fabric-scoped writes has been added to the write testing utility for the data model provider.
* The public interfaces have been moved to `BUILD.gn`. The build files for platforms and examples have been updated.

#### Related issues

#36538

#### Testing

Added unit tests for the new classes and verified they passed:
```
ninja -C out/debug linux_x64_clang/tests/TestOTARequestorCluster && ./out/debug/linux_x64_clang/tests/TestOTARequestorCluster
ninja: Entering directory `out/debug'
[2/2] ld linux_x64_clang/tests/TestOTARequestorCluster
[==========] Running all tests.
[ RUN      ] TestOTARequestorCluster.TestCreate
[       OK ] TestOTARequestorCluster.TestCreate
[ RUN      ] TestOTARequestorCluster.AttributeListTest
[       OK ] TestOTARequestorCluster.AttributeListTest
[ RUN      ] TestOTARequestorCluster.AcceptedCommandsTest
[       OK ] TestOTARequestorCluster.AcceptedCommandsTest
[ RUN      ] TestOTARequestorCluster.GeneratedCommandsTest
[       OK ] TestOTARequestorCluster.GeneratedCommandsTest
[ RUN      ] TestOTARequestorCluster.EventInfoTest
[       OK ] TestOTARequestorCluster.EventInfoTest
[ RUN      ] TestOTARequestorCluster.AnnounceOtaProviderCommandTest
[       OK ] TestOTARequestorCluster.AnnounceOtaProviderCommandTest
[ RUN      ] TestOTARequestorCluster.AnnounceOtaProviderCommandInvalidMetadataTest
[1763065388.416] [1571374:1571374] [ZCL] Metadata size 513 exceeds max 512
[       OK ] TestOTARequestorCluster.AnnounceOtaProviderCommandInvalidMetadataTest
[ RUN      ] TestOTARequestorCluster.ReadAttributesTest
[       OK ] TestOTARequestorCluster.ReadAttributesTest
[ RUN      ] TestOTARequestorCluster.StateTransitionEvent
[       OK ] TestOTARequestorCluster.StateTransitionEvent
[ RUN      ] TestOTARequestorCluster.StateTransitionEventWithNoOtaRequestor
[       OK ] TestOTARequestorCluster.StateTransitionEventWithNoOtaRequestor
[ RUN      ] TestOTARequestorCluster.NoStateTransitionEventWithIdenticalStates
[1763065388.416] [1571374:1571374] [ZCL] Previous state and new state are the same (5), no event to log
[       OK ] TestOTARequestorCluster.NoStateTransitionEventWithIdenticalStates
[ RUN      ] TestOTARequestorCluster.VersionAppliedEvent
[       OK ] TestOTARequestorCluster.VersionAppliedEvent
[ RUN      ] TestOTARequestorCluster.VersionAppliedEventWithNoOtaRequestor
[       OK ] TestOTARequestorCluster.VersionAppliedEventWithNoOtaRequestor
[ RUN      ] TestOTARequestorCluster.DownloadErrorEvents
[       OK ] TestOTARequestorCluster.DownloadErrorEvents
[ RUN      ] TestOTARequestorCluster.DownloadErrorEventsWithNoOtaRequestor
[       OK ] TestOTARequestorCluster.DownloadErrorEventsWithNoOtaRequestor
[ RUN      ] TestOTARequestorCluster.RegistersAsEventHandler
[       OK ] TestOTARequestorCluster.RegistersAsEventHandler
[ RUN      ] TestOTARequestorCluster.ReadsWithNoRequestorInterfaceReturnErrors
[       OK ] TestOTARequestorCluster.ReadsWithNoRequestorInterfaceReturnErrors
[ RUN      ] TestOTARequestorCluster.CommandsWithNoRequestorInterfaceReturnErrors
[       OK ] TestOTARequestorCluster.CommandsWithNoRequestorInterfaceReturnErrors
[ RUN      ] TestOTARequestorCluster.WriteDefaultProvidersList
[       OK ] TestOTARequestorCluster.WriteDefaultProvidersList
[ RUN      ] TestOTARequestorCluster.WriteDefaultProvidersListWithNoInterfaceReturnsErrors
[       OK ] TestOTARequestorCluster.WriteDefaultProvidersListWithNoInterfaceReturnsErrors
[ RUN      ] TestOTARequestorCluster.WritingReadOnlyAttributesReturnsUnsupportedWrite
[       OK ] TestOTARequestorCluster.WritingReadOnlyAttributesReturnsUnsupportedWrite
[==========] Done running all tests.
[  PASSED  ] 21 test(s).
```

Used chip-tool to verify that the cluster responded to attribute reads and the AnnounceOTAProvider command.
```
Terminal 1:
> out/linux-x64-ota-requestor-clang-no-ble/chip-ota-requestor-app --passcode 20202021 --discriminator 3840
Terminal 2:
> out/chip-tool/chip-tool interactive start
>>> pairing code 0x12344321 MT:-24J0AFN00KA0648G00
>>> otasoftwareupdaterequestor read update-possible 0x12344321 0
>>> otasoftwareupdaterequestor announce-otaprovider 0x000000000001B669 1 0 0 0x12344321 0
```

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

harimau-qirex committed:
Add skeleton for code-driven cluster implementation with a couple tests working.

harimau-qirex committed:
Support accepted commands list.

harimau-qirex committed:
Verify generated commands.

harimau-qirex committed:
Add test to verify event info.

harimau-qirex committed:
Add a basic test for the AnnounceOTAProvider command that verifies that it passes the payload to an OTARequestorInterface.

harimau-qirex committed:
Add a test that verifies that AnnounceOTAProvider commands with invalid metadata are handled correctly.

The metadata is invalid if it's too long.

harimau-qirex committed:
Add a test and skeleton for reading attributes.

harimau-qirex committed:
Add support for reading most attributes.

* Store the update state and update state progress attributes in DefaultOTARequestor.
* Remove some initialisation code that's no longer needed due to that.

harimau-qirex committed:
Add support for reading the default OTA providers attribute.

harimau-qirex committed:
Add a default OTA provider to the list during the attribute read test.

harimau-qirex committed:
Expand the read attribute test to verify several update state progress values.

harimau-qirex committed:
Make the cluster handle events.

harimau-qirex committed:
Add an event handler for the OTA requestor interface to call when an event happens.

Add a registry to keep track of event handlers and use it with the OTA requestor cluster.

harimau-qirex committed:
Use the event handler registry in the default OTA requestor implementation.

harimau-qirex committed:
Rename OTA requestor cluster to be consistent with other files ('Ota' -> 'OTA').

harimau-qirex committed:
Rename the class name too.

harimau-qirex committed:
Add start of codegen integration.

harimau-qirex committed:
Attempting to fix build files.

harimau-qirex committed:
Minor fixes

harimau-qirex committed:
Undo the BUILD.gn changes and get back to a working state.

harimau-qirex committed:
Remove the obsolete helpers for DefaultOTARequestor.cpp.

harimau-qirex committed:
Fix the initialization for UpdatePossible.

harimau-qirex committed:
Update the ZAP configuration for the OTA requestor attributes and command.

harimau-qirex committed:
Add the OTA requestor to the code-driven cluster list.

harimau-qirex committed:
Let the OTARequestorInterface be set after initialisation on OTARequestorCluster.

harimau-qirex committed:
Change of approach: move towards a subclass that will check the global singleton.

By default the class will only check what it was given in the constructor.

harimau-qirex committed:
Use a subclass that relies on the global singleton for its data.

harimau-qirex committed:
Add a comment for the method that provides the OTA requestor.

harimau-qirex committed:
Use the newer generated cluster header files.

harimau-qirex committed:
Use the newer generated cluster header files for the new class.

harimau-qirex committed:
Add most of a test for writing the default OTA providers list.

harimau-qirex committed:
Finish the write default OTA providers test.

harimau-qirex committed:
Add a test to verify the behaviour when the OTA requestor isn't set.

harimau-qirex committed:
Write a test to verify the read-only attributes can't be written.

harimau-qirex committed:
Test that the cluster can be created without an OTA requestor too.

harimau-qirex committed:
Fix typos, add some comments.

harimau-qirex committed:
Move the storage for UpdatePossible to the cluster class.

harimau-qirex committed:
Update the codegen version to load and set UpdatePossible.

harimau-qirex committed:
Use some of the helper utility methods in the tests.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

restyled-commits committed:
Restyled by clang-format

harimau-qirex committed:
Fixes for compile issues.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Don't use ClusterTester for commands because it lies about the return value of InvokeCommand.

harimau-qirex committed:
Make the new sources available to CMake.

harimau-qirex committed:
Add a setter/getter for UpdatePossible in case anyone needs it.

harimau-qirex committed:
Start moving source files into BUILD.gn.

harimau-qirex committed:
Update all examples that include the OTA requestor sources directly with CMake.

harimau-qirex committed:
Fix all but one ninja builds that reference the OTA requestor headers.

Not included is nxp/common, as that one is more complicated.

harimau-qirex committed:
Fix some include issues with some platforms.

harimau-qirex committed:
Add missing deps declaration to the shell commands build file.

harimau-qirex committed:
Fix trying to double-set the deps declaration in the shell commands build file, and fix another couple includes.

harimau-qirex committed:
Split the BDX downloader into its own source set. Include it in the ESP32 build.

restyled-commits committed:
Restyled by gn

harimau-qirex committed:
Go back to a simpler set of OTA requestor build file changes.

* Add the new files to BUILD.gn as part of the existing target.
* Move the interface headers to BUILD.gn as a new target.

harimau-qirex committed:
Fix typo in build file.

harimau-qirex committed:
Fix missing references in the ESP32 and nxp platforms. Also silence lint errors.

The optional access errors found by lint aren't issues as the accesses have ASSERT checks immediately beforehand that verify that the optional contains a value.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Add a DefaultOTARequestor to the all clusters app.

harimau-qirex committed:
Don't define the executable path for updates multiple times.

harimau-qirex committed:
Move the image exec path into the image processor class.

harimau-qirex committed:
Correctly initialise kImageExecPath.

harimau-qirex committed:
Add an OTARequestorInterface implementation to the lock app and enable the OTA requestor in unified builds.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Don't enable the OTA requestor in all examples.

harimau-qirex committed:
Make the path of the new executable an actual constant.

Otherwise including this header without referencing kImageExecPath is an unused variable warning in g++.

harimau-qirex committed:
Merge branch 'master' into fix-image-exec-path

harimau-qirex committed:
Revert commits that change the image exec path.

This reverts commits 1cca719eba42b88e107354dbde1e7a19e0672afc,
0e3d938a8b227fa82007f1c6e16b9bb81bf21667, and
89348fa452bfd44137df31e11909c20bd3adc8fe.

Fixing that issue will be done in a different PR.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Merge branch 'master' into fix-image-exec-path

harimau-qirex committed:
Merge branch 'master' into fix-image-exec-path

harimau-qirex committed:
Don't include the OTA requestor cluster in codegen if the build flag isn't set.

harimau-qirex committed:
Merge branch 'fix-image-exec-path' into ota-requestor

harimau-qirex committed:
Don't set up the OTA requestor if the build flag isn't set for the Linux all-clusters app.

harimau-qirex committed:
Fall back to the attribute default values for the OTA requestor in certain circumstances.

Use the fallback when the cluster is in the .zap file but no singleton OTARequestorInstance is set and the flag to enable the OTA requestor is disabled.
Undo the changes to the codegen integration so the OTA requestor cluster is still registered regardless of the flag to enable the OTA requestor.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Handle all CHIP_ERROR return values.

harimau-qirex committed:
Update the tests to handle the new behaviour when the OTA requestor build flag is disabled.

harimau-qirex committed:
Handle a CHIP_ERROR return value in the lock app.

soares-sergio commented on src/app/clusters/ota-requestor/OTARequestorCluster.h:
```diff
@@ -0,0 +1,77 @@
+
+protected:
+    // Returns the OTA requestor instance to use internally. Defaults to the instance passed into the constructor.
+    virtual OTARequestorInterface * OtaRequestorInstance() { return mOtaRequestor; }
```
Could probably just make the member protected and save some flash?

harimau-qirex replied:
This is used in the codegen integration because of initialisation order. The original code used the global singleton, but with codegen the cluster init callbacks are called before the global singleton is set. Making `mOtaRequestor` protected will fail right now because there isn't a clear place to know when the global singelton is set and update the registered clusters in a way transparent to existing codebases.

harimau-qirex commented on src/app/clusters/ota-requestor/OTARequestorCluster.cpp:
```diff
@@ -0,0 +1,252 @@
+        }
+        return encoder.Encode(OtaRequestorInstance()->GetCurrentUpdateStateProgress());
+    case Globals::Attributes::FeatureMap::Id:
+        return encoder.Encode<uint32_t>(0);
```
Correct, the OTA requestor cluster has no features.

harimau-qirex committed:
Improve the comments detailing the behaviour when the OTA requestor build flag is disabled.

harimau-qirex committed:
Expand the OTA requestor README to include how to use OTARequestorCluster.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

soares-sergio commented on src/app/clusters/ota-requestor/BUILD.gn:
```diff
@@ -15,11 +15,32 @@ import("//build_overrides/chip.gni")
+  ]
+
   # TODO: DefaultOTARequestor depends on controller, however the cluster server itself does not.
   # Maybe DefaultOTARequestor and related code should have its own source set.
```
Does DefaultOTARequestor still depends on controller? Can it have its own source set listed here (even if we don't update the apps to use it yet)?

This would at least confirm it builds on its own.

soares-sergio reviewed (approved):
LGTM! Made a few comments but feel free to punt some of those to a follow up, especially the ones that take time to implement.

harimau-qirex committed:
Fix typo.

Co-authored-by: Sergio Soares <sergiosoares@google.com>

restyled-commits committed:
Restyled by clang-format

harimau-qirex committed:
Fix typo in Ameba air purifier definition of OTA requestor sources.

Co-authored-by: Sergio Soares <sergiosoares@google.com>

harimau-qirex committed:
Fix typos in the Ameba chef cmake file.

soares-sergio commented on src/app/clusters/ota-requestor/OTARequestorCluster.h:
@harimau-qirex : could you also run the elf size diff tool and try to look at why it increases code size so much?

Here's how to use the tool (basically build with master on the base same commit for your branch and then build with your branch and diff the elf files):
https://github.com/project-chip/connectedhomeip/blob/4d750897ceb079eb11279980f431fb41eb8b861d/scripts/tools/ELF_SIZE_TOOLING.md#determine-difference-between-two-binaries

harimau-qirex replied:
I checked the realtek lighting app as a first pass. What I found is a lot of the methods in `ota-requestor-server.cpp` were optimised out of the build. All the event handling and most of the attribute reading/writing methods were excluded from the build. I believe this is because the build I looked at defines the OTA requestor but doesn't enable the appropriate build flag, so DefaultOTARequestor is never used and the optimiser removes it and all its dependencies in `ota-requestor-server.cpp`.

While the above accounts for a large chunk of the size difference, the new code is also just bigger. It's difficult to directly compare the changes due to the optimisations I mentioned above, but between overhead (`OTARequestorCluster`'s constructor, destructor, startup, and shutdown; the global singleton callback for `CodegenIntegration.cpp` to work; new init callbacks) and new template instances (`SimpleEventPayloadWriter`, `DataModel::Encode`) there's another large chunk of the size difference.

harimau-qirex committed:
Use aliases for the OTA requestor enums in the event handler.

harimau-qirex committed:
Create a new source set for the default OTA requestor implementation.

This is just step one; it isn't being used anywhere.

harimau-qirex commented on src/app/clusters/ota-requestor/OTARequestorEventHandler.h:
```diff
@@ -0,0 +1,43 @@
+
+class OTARequestorEventHandler
+{
+public:
```
Done.

harimau-qirex committed:
Set the OTA requestor interface used by the cluster objects when the global instance is set.

This is only intended to be used when the codegen data model provider is in use. When using the data model provider the cluster objects should be given the requestor interface when constructed.

harimau-qirex committed:
Remove the now-unnecessary OTA requestor lookup method.

harimau-qirex committed:
Update the event handler registration when changing the OTA requestor.

harimau-qirex committed:
Formatting fix.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Undo creating a new source set.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Update the tests to match the new testing namespace.

harimau-qirex committed:
Remove the cache from the event handler registry.

harimau-qirex committed:
Remove OTARequestorInterface::GetCurrentUpdateStateProgress as it's duplicating an existing method.

harimau-qirex committed:
Remove the event handler registry in favour of a simple array.

We expect that most of the time there will only be one OTA requestor cluster.

andy31415 commented on src/app/clusters/ota-requestor/OTARequestorCluster.cpp:
```diff
@@ -0,0 +1,273 @@
+            // There are examples which enable the OTA requestor cluster in their zap configuration but don't
+            // enable the build flag that controls the implementation of the OTA requestor. The behaviour here
+            // maintains backwards compatibility for those examples.
+#if CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR
```
How about using a different cluster for CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR and in that case only CodegenIntegration picks what clusters to register?

I.e. we register only either a fully valid cluster or the backwards compatibility cluster, but ifdefs stay in codegen integration.

harimau-qirex committed:
Remove the ability to set the OTA requestor.

Codegen uses a fallback to handle attribute reads until the global singleton OTARequestorInterface is set.

This saves space for code driven apps but costs additional space for codegen apps.

harimau-qirex committed:
Remove the event handler interface and have DefaultOTARequestor directly send the events.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Updates to match new APIs.

restyled-commits committed:
Restyled by clang-format

harimau-qirex committed:
Finish removing the event handler references.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
No else-after-return. Specify the boolean encoding for UpdatePossible.

harimau-qirex committed:
Missed some returns inside if-blocks.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Merge commit '3b2cf2b' into ota-requestor

harimau-qirex committed:
Remove obsolete code.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Remove extra error log.

harimau-qirex committed:
Move some code around to reduce diffs from changes.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Merge branch 'master' into ota-requestor

soares-sergio commented on src/app/clusters/ota-requestor/app_config_dependent_sources.cmake:
```suggestion
# Copyright (c) 2025-2026 Project CHIP Authors
```

soares-sergio commented on src/app/clusters/ota-requestor/OTARequestorCluster.h:
```diff
@@ -0,0 +1,58 @@
+
+    CHIP_ERROR AcceptedCommands(const ConcreteClusterPath & path,
+                                ReadOnlyBufferBuilder<DataModel::AcceptedCommandEntry> & builder) override;
+
```
Should we have Get/Set methods here for the readable and writable attributes so the application can configure them?

If not, could you add a comment on why the application would NEVER need to read or write to those attributes.

andy31415 commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,204 @@
+    }
+};
+
+LazyRegisteredServerCluster<OTARequestorCluster> gServers[kOtaRequestorMaxClusterCount];
```
Maybe SetRequestor() can create the class dynamically .... and if that happens, then compiler can LTO instances of OTARequestorCluster if its constructor is never called.

harimau-qirex committed:
Fix copyright year in headers.

harimau-qirex committed:
Simplify the codegen integration since the OTA requestor is a node utility cluster, so only one is allowed per node.

harimau-qirex committed:
Rename the cluster used when the singleton hasn't been set yet.

Also expand the comment explaining its purpose.

harimau-qirex committed:
Verify that the command handler exists so a response to the announce OTA provider command is always sent.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorCluster.cpp:
```diff
@@ -16,116 +16,109 @@
+            CHIP_ERROR error                                = CHIP_NO_ERROR;
+            while (error == CHIP_NO_ERROR && providerIterator.Next())
+            {
+                error = listEncoder.Encode(providerIterator.GetValue());
```
Any reason this is not just ReturnErrorOnFailure() with a simpler loop condition and unconditional `return CHIP_NO_ERROR` after the loop?

harimau-qirex replied:
There isn't a reason, so I've simplified this.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorCluster.cpp:
```diff
@@ -149,156 +142,29 @@ CHIP_ERROR OtaSoftwareUpdateRequestorAttrAccess::WriteDefaultOtaProviders(const
-            ChipLogError(Zcl, "Failed to record DownloadError event: %" CHIP_ERROR_FORMAT, err.Format());
+            ChipLogError(Zcl, "Metadata size %u exceeds max %u", static_cast<unsigned>(metadataForNode.Value().size()),
+                         static_cast<unsigned>(kMaxMetadataLen));
+            return Protocols::InteractionModel::Status::InvalidCommand;
```
Spec says:
```suggestion
            return Protocols::InteractionModel::Status::ConstraintError;
```

harimau-qirex committed:
Add comment explaning how to modify server attributes.

harimau-qirex committed:
Simplify the code to read the default OTA providers list.

harimau-qirex committed:
Return a constraint error when the metadata in the announce command is too large to match spec.

harimau-qirex committed:
Mark the UpdatePossible attribute as changed when it's modified.

harimau-qirex committed:
Add class to deal with attributes.

harimau-qirex committed:
Use an OTARequestorAttributes to store the attributes in DefaultOTARequestor.

harimau-qirex committed:
Add an attributes instance to the codegen integration.

harimau-qirex committed:
Invert the order of the OTARequestorAttributes::Init parameters.

harimau-qirex committed:
Change OTARequestorCluster to use OTARequestorAttributes as the attribute source.

harimau-qirex committed:
Update DefaultOTARequestor and CodegenIntegration with the changes to the attributes.

harimau-qirex committed:
Add a test to verify changed attributes are marked so.

harimau-qirex committed:
Formatting fixes.

harimau-qirex committed:
Add default OTA provider list to the attributes class.

harimau-qirex committed:
Change how to refer to attribute constants.

harimau-qirex committed:
Add storage to the attributes class to handle the default provider list.

harimau-qirex committed:
Rename a method to be clearer.

harimau-qirex committed:
Update OTARequestorCluster to use OTARequestorAttributes for all attributes.

harimau-qirex committed:
Split the command handling call from the rest of the OTARequestorInterface.

harimau-qirex committed:
Change OTARequestorCluster to just use the command interface.

harimau-qirex committed:
Simplify the codegen integration by using a simple forwarder and only one OTARequestorCluster instance.

harimau-qirex committed:
Rename the global OnSetRequestorInstance method.

harimau-qirex committed:
Update a comment.

harimau-qirex committed:
Update attributes test with changes to function name.

harimau-qirex committed:
Add storage-related tests for OTARequestorAttributes.

harimau-qirex committed:
Fix style. Add a guard to verify that the server is constructed in CodegenIntegration.

harimau-qirex committed:
Change the namespace for methods exposed in CodegenIntegration.

harimau-qirex committed:
Update all examples to provide the OTARequestorAttributes from CodegenIntegration.

harimau-qirex committed:
Update the documentation with OTARequestorAttributes.

harimau-qirex commented on src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:
```diff
@@ -659,15 +661,10 @@ IdleStateReason DefaultOTARequestor::MapErrorToIdleStateReason(CHIP_ERROR error)
-        DataModel::Nullable<uint8_t> percent;
-        percent.SetNull();
-        OtaRequestorServerSetUpdateStateProgress(percent);
+        mCurrentUpdateStateProgress.SetNull();
```
I've updated the PR to add dirty-marking for attribute changes. I extracted the attributes into a new class `OTARequestorAttributes` and now `DefaultOTARequestor` and `OTARequestorCluster` depend on that class. That class is responsible for dirty-marking. This adds a new parameter to `DefaultOTARequestor::Init` and I've updated the examples in the SDK where that method is called.

This also let me simplify `CodegenIntegration.cpp` so that only one cluster instance is needed.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Update the README to include OTARequestorAttributes in the how-to section.

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

harimau-qirex committed:
Add the attributes reference to the Linux all-clusters app.

harimau-qirex committed:
Add the new attributes class to the CMake build sources.

harimau-qirex committed:
Update test to use new list writing method for attributes.

harimau-qirex committed:
Update Linux lock app to provide the attributes class.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Re-fix the Linux lock app.

harimau-qirex committed:
Fix description of internal OnSetRequestor callback.

harimau-qirex committed:
Remove misleading file comment.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Formatting after merge.

bzbarsky-apple commented on examples/all-clusters-app/infineon/psoc6/src/AppTask.cpp:
Needs year updated?

harimau-qirex replied:
Done.

bzbarsky-apple commented on examples/all-clusters-app/linux/main-common.cpp:
```diff
@@ -61,6 +69,9 @@
 #include <platform/DeviceInstanceInfoProvider.h>
 #include <platform/DiagnosticDataProvider.h>
+#if CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR
+#include <platform/Linux/OTAImageProcessorImpl.h>
```
This file is compiled on Darwin, but presumably so is platform/Linux/OTAImageProcessorImpl if this works?

harimau-qirex replied:
This has been removed.

bzbarsky-apple commented on examples/all-clusters-app/linux/main-common.cpp:
```diff
@@ -119,6 +130,17 @@ Clusters::ModeSelect::StaticSupportedModesManager sStaticSupportedModesManager;
 Clusters::TimeSynchronization::ExtendedTimeSyncDelegate sTimeSyncDelegate;
 
+#if CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR
+DefaultOTARequestor gRequestorCore;
```
I have to ask: why are we doing all this?  The existing testbed for OTA requestor is the OTA requestor app.  Why are we adding this support to other apps as part of a conversion PR?  This seems like a totally separate piece of work from the actual conversion....

harimau-qirex replied:
A number of the example apps have a partially implemented OTA requestor that doesn't work correctly: the cluster is included in the ZAP file, but the actual OTA requestor cluster code isn't included. I had started fixing it in this PR before I realized the extent of the changes. I've removed these changes from this PR as you recommended, as it distracts from the purpose of this PR.

bzbarsky-apple commented on examples/lock-app/linux/main.cpp:
```diff
@@ -18,6 +18,15 @@
 
 #include "AppMain.h"
 #include <app-common/zap-generated/ids/Clusters.h>
+#if CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR
```
Again, why are we adding this support as part of an "under the hood implementation change" PR?

harimau-qirex replied:
Same as above: this was the start of fixing a number of partial implementations of the OTA requestor that I should've left for a separate PR. I've removed this as well.

bzbarsky-apple commented on examples/platform/ameba/ota/OTAInitializer.cpp:
Copyright year?  Please check all the files for this.

harimau-qirex replied:
Done.

bzbarsky-apple commented on src/app/clusters/ota-requestor/tests/TestOTARequestorAttributes.cpp:
I did not review the tests carefully.

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,151 @@
+// The OTA requestor cluster may be registered by the application before the OTA requestor singleton instance is set.
+// Since the OTA requestor cluster uses the singleton as its data source, this means the cluster won't have meaningful
+// data to return for attributes, nor can it properly handle commands. When this happens, an empty OTA requestor cluster
+// is registered that returns default values. A real cluster object replaces the empty one when the singleton is set.
```
This comment no longer seems to match the code?  The code registers a real cluster, which may have a stubbed out HandleAnnounceOTAProvider but has all the other pieces....

harimau-qirex replied:
I've updated the comment to match the updated code. Thanks for catching this.

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,151 @@
+public:
+    void HandleAnnounceOTAProvider(
+        CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+        const Clusters::OtaSoftwareUpdateRequestor::Commands::AnnounceOTAProvider::DecodableType & commandData) override
```
```suggestion
        const OtaSoftwareUpdateRequestor::Commands::AnnounceOTAProvider::DecodableType & commandData) override
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:
```diff
@@ -96,45 +98,37 @@ static void LogApplyUpdateResponse(const ApplyUpdateResponse::DecodableType & re
+        return;
+    }
     globalOTARequestorInstance = instance;
+    if (gInternalOnSetRequestorInstance != nullptr)
```
Wait, but then how do you remove a reference to an instance that is going away?

harimau-qirex replied:
This is checking that the callback function isn't null. The callback will still get called when the `OTARequestorInstance` is null.

bzbarsky-apple replied:
Right you are.

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:
```diff
@@ -581,39 +589,32 @@ void DefaultOTARequestor::NotifyUpdateApplied()
+        {
+            if (cluster.clusterId == OtaSoftwareUpdateRequestor::Id)
+            {
+                static_cast<DataModel::EventsGenerator &>(EventManagement::GetInstance()).GenerateEvent(event, endpoint.id);
```
Why are we explicitly using `EventManagement::GetInstance()`?  Shouldn't we be getting the events generator from the provider and/or the IM engine?

Same in other places below.

harimau-qirex replied:
It doesn't look like there's a way to get the events generator from the provider or the IM engine. It would be best to have `OTARequestorCluster` to provide the events generator, but that's difficult due to initialization order. I've created an interface to proxy events from `DefaultOTARequestor` to `OTARequestorCluster` and added it to `DefaultOTARequestor::Init` in the same way as `OTARequestorAttributes`. This also removes the need to look up what endpoint contains the OTA requestor cluster.

harimau-qirex committed:
Undo the changes to the Linux apps that added the OTA requestor implementation.

Fixing the apps should be deferred to another PR.

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:
```diff
@@ -677,11 +673,28 @@ void DefaultOTARequestor::RecordNewUpdateState(OTAUpdateStateEnum newState, OTAC
-    mCurrentUpdateState = newState;
+    mAttributes->SetUpdateState(newState);
+
+    if (prevState != newState)
```
Shouldn't this event generation be handled by the attributes?  Are there any use cases for changing the state in the attributes and _not_ generating the event?

harimau-qirex replied:
The only use case for that is when loading the update state from storage on boot. I've update `OTARequestorAttributes` so it handles that case when the storage is set and moved the event generation to that class as well.

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:
```diff
@@ -928,10 +956,9 @@ void DefaultOTARequestor::LoadCurrentUpdateInfo()
-        mCurrentUpdateState = OTAUpdateStateEnum::kIdle;
-    }
+    OTAUpdateStateEnum updateState = OTAUpdateStateEnum::kIdle;
+    TEMPORARY_RETURN_IGNORED mStorage->LoadCurrentUpdateState(updateState);
```
Why the behavior change from just ignoring the possibly-garbage value to using it?

harimau-qirex replied:
I mistakenly assumed that LoadCurrentUpdateState wasn't allowed to change the value unless it returned CHIP_NO_ERROR. I've undone the behaviour change.

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestor.h:
```diff
@@ -100,7 +104,10 @@ class DefaultOTARequestor : public OTARequestorInterface, public BDXDownloader::
-    ProviderLocationList::Iterator GetDefaultOTAProviderListIterator(void) override { return mDefaultOtaProviderList.Begin(); }
+    ProviderLocationList::Iterator GetDefaultOTAProviderListIterator(void) override
+    {
+        return mAttributes->GetDefaultOtaProviderListIterator();
```
Why can we assume mAttributes is non-null here, but not in GetCurrentUpdateState?

harimau-qirex replied:
We can't. Fixed.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.cpp:
```diff
@@ -0,0 +1,136 @@
+
+constexpr ClusterId kClusterId = app::Clusters::OtaSoftwareUpdateRequestor::Id;
+
+void OTARequestorAttributes::SetUpdateState(OTAUpdateStateEnum updateState)
```
All of the functions in this file should have assertChipStackLockedByCurrentThread() at the top, since none of this stuff is threadsafe, right?

harimau-qirex replied:
Done.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.cpp:
```diff
@@ -0,0 +1,136 @@
+
+    // Ignore the error if no entry for the associated fabric index has been found.
+    VerifyOrReturnError(error != CHIP_ERROR_NOT_FOUND, CHIP_NO_ERROR);
+    ReturnErrorOnFailure(error);
```
```suggestion
    ReturnErrorOnFailure(error.NoErrorIf(CHIP_ERROR_NOT_FOUND));
```

and then the comment might be superfluous, not sure.

harimau-qirex replied:
This would still let the rest of the function run which would mark the default providers list as dirty and store the default providers list. I've reworded the comment and changed the return statement to make it clearer what behaviour we want.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.cpp:
```diff
@@ -0,0 +1,136 @@
+        ProviderLocationType location = iterator.GetValue();
+        if (location.GetFabricIndex() == providerLocation.GetFabricIndex())
+        {
+            ChipLogError(SoftwareUpdate, "Default OTA provider entry with fabric %d already exists", location.GetFabricIndex());
```
```suggestion
            ChipLogError(SoftwareUpdate, "Default OTA provider entry with fabric index %d already exists", location.GetFabricIndex());
```

harimau-qirex replied:
Done.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.h:
```diff
@@ -0,0 +1,73 @@
+
+// Class that tracks the non-array attributes in the OTA Requestor cluster. Marks
+// attributes dirty when they're changed. Updates the storage when the default OTA
+// providers list changes.
```
Should document the expectation that all methods are invoked with the Matter lock held, right?

harimau-qirex replied:
Done.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.h:
```diff
@@ -0,0 +1,73 @@
+    using OTAUpdateStateEnum   = OTARequestorInterface::OTAUpdateStateEnum;
+    using ProviderLocationType = OTARequestorInterface::ProviderLocationType;
+
+    OTAUpdateStateEnum GetUpdateState() const { return mUpdateState; }
```
These inlined methods should assertChipStackLockedByCurrentThread too (and maybe not be inline).

harimau-qirex replied:
Done.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.h:
```diff
@@ -0,0 +1,73 @@
+
+    CHIP_ERROR AddDefaultOtaProvider(const ProviderLocationType & providerLocation);
+    CHIP_ERROR RemoveDefaultOtaProvider(FabricIndex fabricIndex);
+    ProviderLocationList::Iterator GetDefaultOtaProviderListIterator();
```
Needs to document whether the iterator is safe to use across add/remove calls (probably not, right?).

harimau-qirex replied:
Done. (no, it isn't)

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorCluster.cpp:
```diff
@@ -149,156 +140,31 @@ CHIP_ERROR OtaSoftwareUpdateRequestorAttrAccess::WriteDefaultOtaProviders(const
+                         static_cast<unsigned>(kMaxMetadataLen));
+            return Protocols::InteractionModel::Status::ConstraintError;
         }
+        mOtaCommands.HandleAnnounceOTAProvider(handler, request.path, data);
```
This also needs to check that announcementReason does not have the unknown enum value, right?

I know the old command handler did not.  Followup is fine for this change.

harimau-qirex replied:
It's sort of handled in that `DefaultOTARequestorDriver::ProcessAnnounceOTAProviders` logs an error and ignores the announcement, but `DefaultOTARequestor::HandleAnnoucneOTAProvider` still marks the command as successfully handled, which is incorrect as far as I'm aware. I'll change this in a followup so that the invalid reason is handled here and better.

bzbarsky-apple replied:
My point is that not checking it here is a spec violation.

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorInterface.h:
```diff
@@ -226,4 +234,11 @@ void SetRequestorInstance(OTARequestorInterface * instance);
+// instance can be set before or after when the OTARequestorCluster instance in CodegenIntegration is created. The callback is used
+// to ensure that regardless of the order the OTARequestorCluster instance can forward commands to the OTARequestorInterface
+// implementation. The implementation of SetRequestorInstance must call this (if not null) after setting the requestor instance.
+// THIS MUST ONLY BE SET IN CodegenIntegration.cpp!
```
Shouldn't this be in a separate header included from the two relevant .cpp files and NOT listed as an exported header in gn for this source set and whatnot (basically have it in a separate group, which is in deps but not public deps for whatever cpp files need it).

harimau-qirex replied:
I've moved this into a separate header and created a source set for it, but cleaning up the build files for the OTA requestor will require additional work that I plan to look at in a separate PR.

harimau-qirex committed:
Update the copyright year on the examples I modified.

harimau-qirex committed:
Update a comment to match how OTA requestor commands are handled with the code-driven data model.

harimau-qirex committed:
Remove unnecessary namespace.

harimau-qirex committed:
Undo change that ignored a return value.

harimau-qirex committed:
Also load the update state attribute when storage is set.

harimau-qirex committed:
Pass an event generator to the attributes and use it when setting the update state.

harimau-qirex committed:
Reword the case when removing a non-existing default provider to make it clearer.

harimau-qirex committed:
Fix a log: "fabric" -> "fabric index"

harimau-qirex committed:
Don't assume the attributes object has been set when getting the default OTA providers list.

harimau-qirex committed:
Explain that the OTA provider list iterator is invalid after calls to add/remove providers.

harimau-qirex committed:
Ensure all methods in OTARequestorAttributes are called while holding the Matter lock.

harimau-qirex committed:
Move the internal callback needed for codegen integration to its own header.

harimau-qirex committed:
Fix the name of the new header.

harimau-qirex committed:
Add an event sender interface used by the DefaultOTARequestor.

Have OTARequestorCluster implement that interface.

harimau-qirex committed:
Expose the OTARequestorCluster instance as the event sender in the codegen integration.

harimau-qirex committed:
Update all calls to DefaultOTARequestor::Init to include the new parameter.

harimau-qirex committed:
Merge branch 'master' into ota-requestor

harimau-qirex committed:
Add another interface for OTARequestorAttributes to notify of attribute changes.

This ensures that the cluster's data version is updated when an attribute changes.

harimau-qirex committed:
Verify that the data version changes (or doesn't) in the OTARequestorCluster tests.

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

bzbarsky-apple commented on src/app/clusters/ota-requestor/tests/TestOTARequestorCluster.cpp:
```diff
@@ -366,30 +376,127 @@ TEST_F(TestOTARequestorCluster, ChangingAttributesMarksAsChanged)
     // DefaultOTAProviders is verified in the test WriteDefaultProvidersList.
 }
 
+TEST_F(TestOTARequestorCluster, SendVersionAppliedEventSendsAnEvent)
```
Events are "generated", not "sent".  Generating an event does not mean that it will be "sent" anywhere; that's a completely separate thing.

This test is testing whether events are generated.

harimau-qirex replied:
I've changed all references from sending events to generating events.

bzbarsky-apple commented on src/app/clusters/ota-requestor/tests/TestOTARequestorCluster.cpp:
```diff
@@ -366,30 +376,127 @@ TEST_F(TestOTARequestorCluster, ChangingAttributesMarksAsChanged)
+
+    VersionApplied::Type event{ 0x12345678u, 0xABCDu };
+    EXPECT_EQ(cluster.SendVersionAppliedEvent(event), CHIP_NO_ERROR);
+    auto sentEvent = eventsGenerator.GetNextEvent();
```
generatedEvent, etc. Please fix all the naming in this file.

bzbarsky-apple commented on src/app/clusters/ota-requestor/app_config_dependent_sources.gni:
```diff
@@ -16,6 +16,7 @@ app_config_dependent_sources = [
   "BDXDownloader.h",
   "CodegenIntegration.cpp",
   "CodegenIntegration.h",
+  "CodegenIntegrationInternal.h",
```
That header is not app config dependent, is it?  But I guess it's only used from things that are?

harimau-qirex replied:
That's correct. The global variable it exposes is only written to when the codegen data model is in use, and the name is to discourage using it when using the codedriven data model. If there's a way to guarantee that the file is only included when using the codegen data model I would do that instead, but I'm unaware of a way to do that.

bzbarsky-apple commented on src/app/clusters/ota-requestor/BUILD.gn:
```diff
@@ -43,3 +43,9 @@ source_set("interface") {
   ]
 }
+
+source_set("codegen-integration-internal") {
```
I don't understand the point of this source set... it's not used, is it? We're just throwing the header into the codegen dependent sources.

I suspect you can remove this for now, but do a followup to fix deps so that the header is only allowed to be included if you go out of your way to include it?  That probably needs a way for the gni to add things to the deps (not public_deps) of app_config_dependent_sources or something via setting some variable and whatever processes app_config_dependent_sources handling that variable.

harimau-qirex replied:
I've removed it and will fix it in a followup.

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,187 @@
+    OTARequestorCommandInterface * mDestination = nullptr;
+};
+
+// This class decouples the cluster and the DefaultOTARequestor singleton's event sending to allow the OTARequestorCluster to be
```
```suggestion
// This class decouples the cluster and the DefaultOTARequestor singleton's event generation to allow the OTARequestorCluster to be
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,187 @@
+
+// This class decouples the cluster and the DefaultOTARequestor singleton's event sending to allow the OTARequestorCluster to be
+// created after DefaultOTARequestor::Init is called.
+class OTAEventForwarder : public DefaultOTARequestorEventSender
```
```suggestion
class OTAEventForwarder : public DefaultOTARequestorEventGenerator
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,187 @@
+public:
+    ~OTAEventForwarder() override = default;
+
+    CHIP_ERROR SendVersionAppliedEvent(const VersionAppliedEvent & event) override
```
```suggestion
    CHIP_ERROR GenerateVersionAppliedEvent(const VersionAppliedEvent & event) override
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,187 @@
+    CHIP_ERROR SendVersionAppliedEvent(const VersionAppliedEvent & event) override
+    {
+        VerifyOrReturnError(mDestination, CHIP_ERROR_INCORRECT_STATE);
+        return mDestination->SendVersionAppliedEvent(event);
```
```suggestion
        return mDestination->GenerateVersionAppliedEvent(event);
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,187 @@
+    CHIP_ERROR SendDownloadErrorEvent(const DownloadErrorEvent & event) override
+    {
+        VerifyOrReturnError(mDestination, CHIP_ERROR_INCORRECT_STATE);
+        return mDestination->SendDownloadErrorEvent(event);
```
```suggestion
    CHIP_ERROR GenerateDownloadErrorEvent(const DownloadErrorEvent & event) override
    {
        VerifyOrReturnError(mDestination, CHIP_ERROR_INCORRECT_STATE);
        return mDestination->GenerateDownloadErrorEvent(event);
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,187 @@
+    return gAttributes;
+}
+
+DefaultOTARequestorEventSender & GetDefaultOTARequestorEventSender()
```
```suggestion
DefaultOTARequestorEventSender & GetDefaultOTARequestorEventGenerator()
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.h:
```diff
@@ -0,0 +1,32 @@
+
+#pragma once
+
+#include <app/clusters/ota-requestor/DefaultOTARequestorEventSender.h>
```
```suggestion
#include <app/clusters/ota-requestor/DefaultOTARequestorEventGenerator.h>
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/CodegenIntegration.h:
```diff
@@ -0,0 +1,32 @@
+OTARequestorAttributes & GetOTARequestorAttributes();
+
+// Returns an event sender that can be used by a DefaultOTARequestor.
+DefaultOTARequestorEventSender & GetDefaultOTARequestorEventSender();
```
I'm going to stop commenting on this "sender" naming, but please fix it throughout.

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:
```diff
@@ -96,45 +99,39 @@ static void LogApplyUpdateResponse(const ApplyUpdateResponse::DecodableType & re
     mOtaRequestorDriver = &driver;
     mBdxDownloader      = &downloader;
+    mAttributes         = &attributes;
+    mEventSender        = &eventSender;
```
```suggestion
    mEventGenerator        = &eventGenerator;
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:
```diff
@@ -581,39 +592,24 @@ void DefaultOTARequestor::NotifyUpdateApplied()
+    {
+        DefaultOTARequestorEventSender::VersionAppliedEvent event{ mCurrentVersion, productId };
+        CHIP_ERROR error = mEventSender->SendVersionAppliedEvent(event);
+        SuccessOrLog(error, SoftwareUpdate, "Failed to record VersionApplied event: %" CHIP_ERROR_FORMAT, error.Format());
```
```suggestion
        SuccessOrLog(error, SoftwareUpdate, "Failed to generate VersionApplied event: %" CHIP_ERROR_FORMAT, error.Format());
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/DefaultOTARequestorEventSender.h:
```diff
@@ -0,0 +1,40 @@
+
+namespace chip {
+
+// This interface is used by the DefaultOTARequestor to send events.
```
```suggestion
// This interface is used by the DefaultOTARequestor to generate events.
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.h:
```diff
@@ -0,0 +1,89 @@
+
+// Class that tracks the non-array attributes in the OTA Requestor cluster. Marks
+// attributes dirty when they're changed. Updates the storage when the default OTA
+// providers list changes. Sends events when the update state changes. All methods
```
```suggestion
// providers list changes. Generates events when the update state changes. All methods
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.h:
```diff
@@ -0,0 +1,89 @@
+    };
+
+    OTAUpdateStateEnum GetUpdateState() const;
+    // If the events generator is set this will also send a StateTransition event.
```
```suggestion
    // If the events generator is set this will also generate a StateTransition event.
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorAttributes.h:
```diff
@@ -0,0 +1,89 @@
+
+    CHIP_ERROR SetInteractionModelContext(EndpointId endpointId, AttributeChangeListener & attributeChangeListener,
+                                          app::DataModel::EventsGenerator & eventsGenerator);
+    // This loads the default OTA provider list and update state attributes from storage. This will not send events.
```
```suggestion
    // This loads the default OTA provider list and update state attributes from storage. This will not generate events.
```

bzbarsky-apple commented on src/app/clusters/ota-requestor/OTARequestorCluster.cpp:
```diff
@@ -149,156 +140,47 @@ CHIP_ERROR OtaSoftwareUpdateRequestorAttrAccess::WriteDefaultOtaProviders(const
+                         static_cast<unsigned>(kMaxMetadataLen));
+            return Protocols::InteractionModel::Status::ConstraintError;
         }
+        mOtaCommands.HandleAnnounceOTAProvider(handler, request.path, data);
```
Please make sure an issue is filed tracking the spec-required validity check on announcementReason.

harimau-qirex replied:
I've filed #71647.

harimau-qirex committed:
Fix EFR32 and BouffaloLab builds.

harimau-qirex committed:
Change mentions of sending events to generating events, since that's what the logic is doing.

harimau-qirex committed:
Update the DefaultOTARequestor::Init calls.

harimau-qirex committed:
Remove useless source set. Will need to fix the OTA requestor's source sets in a future PR.

harimau-qirex committed:
Missed a few references to sending events.

harimau-qirex committed:
Update the cmake file too.

soares-sergio commented:
In a follow up, let's add this to all-devices-app so it can be OTA'd. That would give us good confidence on the new model.

harimau-qirex committed:
Merge branch 'master' into ota-requestor
